Crate is not actually `no_std`

Despite the assertion in the docs, the `vek` crate at 0.9.6 does not actually build on targets without `std` available, because of a transitive dependency on `std` through the `approx` crate.

# Steps to reproduce:

Ensure that the `thumbv7em-none-eabihf` target is installed:

```shell
$ rustup target add thumbv7em-none-eabihf
```

Clone and build:

```shell
$ git clone https://github.com/yoanlcq/vek.git
$ cd vek
$ cargo build --target thumbv7em-none-eabihf
```

# Expected results

Crate builds.

# Actual results

```
    Updating crates.io index
   Compiling semver-parser v0.7.0
   Compiling num-traits v0.2.6
   Compiling num-integer v0.1.39
   Compiling approx v0.1.1
   Compiling static_assertions v0.2.5
error[E0463]: can't find crate for `std`
  |
  = note: the `thumbv7em-none-eabihf` target may not be installed

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: Could not compile `approx`.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

# Likely fix

`approx`, `num-traits`, `num-integer` all need to be set to `default-features = false`. Unfortunately, the version of `num-integer` (at least) that's pinned here isn't `no_std` compatible, so this probably never worked. Upgrading to at least 0.1.36 should fix it.

Also, consider adding a no-std target to your continuous build to catch regressions here. `thumbv7em-none-eabihf` is a supported target on stable and nightly and would be a reasonable choice, but in general, any target with `-none-` in the triple instead of e.g. `-linux-` should do.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -22,7 +22,7 @@ appveyor  = { repository = "yoanlcq/vek" }
 maintenance = { status = "experimental" }
 
 [features]
-default = ["repr_simd", "rgba", "rgb"]
+default = ["repr_simd", "rgba", "rgb", "std"]
 # NOTE: The below ones (commented out) are enforced, because doc tests assume their availability.
 # vec2 = []
 # vec3 = []
@@ -44,14 +44,16 @@ rgba = []
 uv = []
 uvw = []
 repr_simd = []
+std = ["num-traits/std"]
+libm = ["num-traits/libm"]
 
 [build-dependencies]
 rustc_version = "~0.2.1"
 
 [dependencies]
-approx = "~0.1.1"
-num-traits = "~0.1.40"
-num-integer = "~0.1.35"
+approx = { version = "0.3.2", default-features = false }
+num-traits = { version = "0.2.10", default-features = false }
+num-integer = { version = "~0.1.41", default-features = false }
 static_assertions = "~0.2.5"
 image = { version = "~0.17", optional = true, default-features = false }
 serde = { version = "~1.0.24", optional = true, features = ["derive"] }

--- a/src/mat.rs
+++ b/src/mat.rs
@@ -7,7 +7,7 @@ use std::iter::Sum;
 use std::fmt::{self, Display, Formatter, Debug};
 use std::ops::*;
 use num_traits::{Zero, One, real::Real, FloatConst, NumCast};
-use approx::ApproxEq;
+use approx::{AbsDiffEq, RelativeEq, UlpsEq};
 use ops::MulAdd;
 use vec;
 use geom::{Rect, FrustumPlanes}; // NOTE: Rect is therefore always repr_c here
@@ -321,7 +321,7 @@ macro_rules! mat_impl_mat {
             ///
             /// # Panics
             /// Panics if the matrix's elements are not tightly packed in memory,
-            /// which may be the case for matrices in the `repr_simd` module.  
+            /// which may be the case for matrices in the `repr_simd` module.
             /// You may check this with the `is_packed()` method.
             pub fn as_row_ptr(&self) -> *const T {
                 assert!(self.is_packed());
@@ -331,7 +331,7 @@ macro_rules! mat_impl_mat {
             ///
             /// # Panics
             /// Panics if the matrix's elements are not tightly packed in memory,
-            /// which may be the case for matrices in the `repr_simd` module.  
+            /// which may be the case for matrices in the `repr_simd` module.
             /// You may check this with the `is_packed()` method.
             pub fn as_mut_row_ptr(&mut self) -> *mut T {
                 assert!(self.is_packed());
@@ -341,7 +341,7 @@ macro_rules! mat_impl_mat {
             ///
             /// # Panics
             /// Panics if the matrix's elements are not tightly packed in memory,
-            /// which may be the case for matrices in the `repr_simd` module.  
+            /// which may be the case for matrices in the `repr_simd` module.
             /// You may check this with the `is_packed()` method.
             pub fn as_row_slice(&self) -> &[T] {
                 unsafe {
@@ -352,7 +352,7 @@ macro_rules! mat_impl_mat {
             ///
             /// # Panics
             /// Panics if the matrix's elements are not tightly packed in memory,
-            /// which may be the case for matrices in the `repr_simd` module.  
+            /// which may be the case for matrices in the `repr_simd` module.
             /// You may check this with the `is_packed()` method.
             pub fn as_mut_row_slice(&mut self) -> &mut [T] {
                 unsafe {
@@ -362,7 +362,7 @@ macro_rules! mat_impl_mat {
         }
 
 
-        /// Displays this matrix using the following format:  
+        /// Displays this matrix using the following format:
         ///
         /// (`i` being the number of rows and `j` the number of columns)
         ///
@@ -372,7 +372,7 @@ macro_rules! mat_impl_mat {
         ///   mi0 ... mij )
         /// ```
         ///
-        /// Note that elements are not comma-separated.  
+        /// Note that elements are not comma-separated.
         /// This format doesn't depend on the matrix's storage layout.
         impl<T: Display> Display for $Mat<T> {
             fn fmt(&self, f: &mut Formatter) -> fmt::Result {
@@ -504,7 +504,7 @@ macro_rules! mat_impl_mat {
                 }
             }
         }
-        /// Multiplies a row-major matrix with a column-major matrix, 
+        /// Multiplies a row-major matrix with a column-major matrix,
         /// producing a column-major matrix.
         ///
         /// ```
@@ -841,7 +841,7 @@ macro_rules! mat_impl_mat {
             ///
             /// # Panics
             /// Panics if the matrix's elements are not tightly packed in memory,
-            /// which may be the case for matrices in the `repr_simd` module.  
+            /// which may be the case for matrices in the `repr_simd` module.
             /// You may check this with the `is_packed()` method.
             pub fn as_col_ptr(&self) -> *const T {
                 assert!(self.is_packed());
@@ -851,7 +851,7 @@ macro_rules! mat_impl_mat {
             ///
             /// # Panics
             /// Panics if the matrix's elements are not tightly packed in memory,
-            /// which may be the case for matrices in the `repr_simd` module.  
+            /// which may be the case for matrices in the `repr_simd` module.
             /// You may check this with the `is_packed()` method.
             pub fn as_mut_col_ptr(&mut self) -> *mut T {
                 assert!(self.is_packed());
@@ -861,7 +861,7 @@ macro_rules! mat_impl_mat {
             ///
             /// # Panics
             /// Panics if the matrix's elements are not tightly packed in memory,
-            /// which may be the case for matrices in the `repr_simd` module.  
+            /// which may be the case for matrices in the `repr_simd` module.
             /// You may check this with the `is_packed()` method.
             pub fn as_col_slice(&self) -> &[T] {
                 unsafe {
@@ -872,7 +872,7 @@ macro_rules! mat_impl_mat {
             ///
             /// # Panics
             /// Panics if the matrix's elements are not tightly packed in memory,
-            /// which may be the case for matrices in the `repr_simd` module.  
+            /// which may be the case for matrices in the `repr_simd` module.
             /// You may check this with the `is_packed()` method.
             pub fn as_mut_col_slice(&mut self) -> &mut [T] {
                 unsafe {
@@ -882,7 +882,7 @@ macro_rules! mat_impl_mat {
         }
 
 
-        /// Displays this matrix using the following format:  
+        /// Displays this matrix using the following format:
         ///
         /// (`i` being the number of rows and `j` the number of columns)
         ///
@@ -892,7 +892,7 @@ macro_rules! mat_impl_mat {
         ///   mi0 ... mij )
         /// ```
         ///
-        /// Note that elements are not comma-separated.  
+        /// Note that elements are not comma-separated.
         /// This format doesn't depend on the matrix's storage layout.
         impl<T: Display> Display for $Mat<T> {
             fn fmt(&self, f: &mut Formatter) -> fmt::Result {
@@ -1233,7 +1233,7 @@ macro_rules! mat_impl_mat {
             pub const COL_COUNT: usize = $ncols;
 
             /// Are all elements of this matrix tightly packed together in memory ?
-            /// 
+            ///
             /// This might not be the case for matrices in the `repr_simd` module
             /// (it depends on the target architecture).
             pub fn is_packed(&self) -> bool {
@@ -1242,7 +1242,7 @@ macro_rules! mat_impl_mat {
             }
         }
 
-        impl<T> Mul<T> for $Mat<T> 
+        impl<T> Mul<T> for $Mat<T>
             where T: Copy + Zero + Add<Output=T> + Mul<Output=T>
         {
             type Output = Self;
@@ -1257,13 +1257,13 @@ macro_rules! mat_impl_mat {
 
         impl<T> MulAssign for $Mat<T>
             where T: Copy + Zero + Add<Output=T> + Mul<Output=T> + MulAdd<T,T,Output=T>
-        { 
+        {
             fn mul_assign(&mut self, rhs: Self) { *self = *self * rhs; }
         }
 
         impl<T> MulAssign<T> for $Mat<T>
             where T: Copy + Zero + Add<Output=T> + Mul<Output=T>
-        { 
+        {
             fn mul_assign(&mut self, rhs: T) { *self = *self * rhs; }
         }
 
@@ -1369,33 +1369,46 @@ macro_rules! mat_impl_mat {
         impl<T: Rem<Output=T> + Copy> RemAssign    for $Mat<T> { fn rem_assign(&mut self, rhs: Self) { *self = *self % rhs; } }
         impl<T: Rem<Output=T> + Copy> RemAssign<T> for $Mat<T> { fn rem_assign(&mut self, rhs: T   ) { *self = *self % rhs; } }
 
-        impl<T: ApproxEq> ApproxEq for $Mat<T> where T::Epsilon: Copy {
+        impl<T: AbsDiffEq> AbsDiffEq for $Mat<T> where T::Epsilon: Copy {
             type Epsilon = T::Epsilon;
 
             fn default_epsilon() -> T::Epsilon {
                 T::default_epsilon()
             }
 
-            fn default_max_relative() -> T::Epsilon {
-                T::default_max_relative()
-            }
-
-            fn default_max_ulps() -> u32 {
-                T::default_max_ulps()
-            }
-
-            fn relative_eq(&self, other: &Self, epsilon: T::Epsilon, max_relative: T::Epsilon) -> bool {
+            fn abs_diff_eq(&self, other: &Self, epsilon: Self::Epsilon) -> bool {
                 for (l, r) in self.$lines.iter().zip(other.$lines.iter()) {
-                    if !ApproxEq::relative_eq(l, r, epsilon, max_relative) {
+                    if !AbsDiffEq::abs_diff_eq(l, r, epsilon) {
                         return false;
                     }
                 }
                 true
             }
+        }
+
+        impl<T: UlpsEq> UlpsEq for $Mat<T> where T::Epsilon: Copy {
+            fn default_max_ulps() -> u32 {
+                T::default_max_ulps()
+            }
 
             fn ulps_eq(&self, other: &Self, epsilon: T::Epsilon, max_ulps: u32) -> bool {
                 for (l, r) in self.$lines.iter().zip(other.$lines.iter()) {
-                    if !ApproxEq::ulps_eq(l, r, epsilon, max_ulps) {
+                    if !UlpsEq::ulps_eq(l, r, epsilon, max_ulps) {
+                        return false;
+                    }
+                }
+                true
+            }
+        }
+
+        impl<T: RelativeEq> RelativeEq for $Mat<T> where T::Epsilon: Copy {
+            fn default_max_relative() -> T::Epsilon {
+                T::default_max_relative()
+            }
+
+            fn relative_eq(&self, other: &Self, epsilon: T::Epsilon, max_relative: T::Epsilon) -> bool {
+                for (l, r) in self.$lines.iter().zip(other.$lines.iter()) {
+                    if !RelativeEq::relative_eq(l, r, epsilon, max_relative) {
                         return false;
                     }
                 }
@@ -1748,7 +1761,7 @@ macro_rules! mat_impl_mat4 {
             /// matrix which scale is 1.
             ///
             /// A transform matrix is invertible this way as long as it consists
-            /// of translations, rotations, and shears.  
+            /// of translations, rotations, and shears.
             /// **It's not guaranteed to work if the scale is not 1.**
             ///
             /// ```
@@ -2001,7 +2014,7 @@ macro_rules! mat_impl_mat4 {
                 *self = self.rotated_x(angle_radians);
             }
             /// Returns this matrix rotated around the X axis.
-            pub fn rotated_x(self, angle_radians: T) -> Self 
+            pub fn rotated_x(self, angle_radians: T) -> Self
                 where T: Real + MulAdd<T,T,Output=T>
             {
                 Self::rotation_x(angle_radians) * self
@@ -2063,21 +2076,21 @@ macro_rules! mat_impl_mat4 {
                     T::zero(), T::zero(), T::zero(), T::one()
                 )
             }
-            /// Rotates this matrix around a 3D axis.  
+            /// Rotates this matrix around a 3D axis.
             /// The axis is not required to be normalized.
             pub fn rotate_3d<V: Into<Vec3<T>>>(&mut self, angle_radians: T, axis: V)
                 where T: Real + MulAdd<T,T,Output=T> + Sum
             {
                 *self = self.rotated_3d(angle_radians, axis);
             }
-            /// Returns this matrix rotated around a 3D axis.  
+            /// Returns this matrix rotated around a 3D axis.
             /// The axis is not required to be normalized.
             pub fn rotated_3d<V: Into<Vec3<T>>>(self, angle_radians: T, axis: V) -> Self
                 where T: Real + MulAdd<T,T,Output=T> + Sum
             {
                 Self::rotation_3d(angle_radians, axis) * self
             }
-            /// Creates a matrix that rotates around a 3D axis.  
+            /// Creates a matrix that rotates around a 3D axis.
             /// The axis is not required to be normalized.
             ///
             /// ```
@@ -2154,7 +2167,7 @@ macro_rules! mat_impl_mat4 {
             /// assert_relative_eq!(m * from, to);
             /// # }
             /// ```
-            pub fn rotation_from_to_3d<V: Into<Vec3<T>>>(from: V, to: V) -> Self 
+            pub fn rotation_from_to_3d<V: Into<Vec3<T>>>(from: V, to: V) -> Self
                 where T: Real + Sum
             {
                 Self::from(Quaternion::rotation_from_to_3d(from, to))
@@ -2167,10 +2180,10 @@ macro_rules! mat_impl_mat4 {
             /// Builds a change of basis matrix that transforms points and directions from
             /// any space to the canonical one.
             ///
-            /// `origin` is the origin of the child space.  
+            /// `origin` is the origin of the child space.
             /// `i`, `j` and `k` are all required to be normalized;
             /// They are the unit basis vector along the target space x-axis, y-axis and z-axis
-            /// respectively, expressed in canonical-space coordinates.  
+            /// respectively, expressed in canonical-space coordinates.
             ///
             /// ```
             /// # extern crate vek;
@@ -2228,10 +2241,10 @@ macro_rules! mat_impl_mat4 {
             /// Builds a change of basis matrix that transforms points and directions from
             /// canonical space to another space.
             ///
-            /// `origin` is the origin of the child space.  
+            /// `origin` is the origin of the child space.
             /// `i`, `j` and `k` are all required to be normalized;
             /// They are the unit basis vector along the target space x-axis, y-axis and z-axis
-            /// respectively, expressed in canonical-space coordinates.  
+            /// respectively, expressed in canonical-space coordinates.
             ///
             /// ```
             /// # extern crate vek;
@@ -2306,7 +2319,7 @@ macro_rules! mat_impl_mat4 {
             /// from an eye position, a target position, and up vector.
             /// Commonly used for cameras - in short, it maps points from
             /// world-space to eye-space.
-            /// 
+            ///
             /// ```
             /// # extern crate vek;
             /// # #[macro_use] extern crate approx;
@@ -2339,7 +2352,7 @@ macro_rules! mat_impl_mat4 {
             /// from an eye position, a target position, and up vector.
             /// Commonly used for cameras - in short, it maps points from
             /// world-space to eye-space.
-            /// 
+            ///
             /// ```
             /// # extern crate vek;
             /// # #[macro_use] extern crate approx;
@@ -2381,7 +2394,7 @@ macro_rules! mat_impl_mat4 {
             /// Builds a "look at" model transform for left-handed spaces
             /// from an eye position, a target position, and up vector.
             /// Preferred for transforming objects.
-            /// 
+            ///
             /// ```
             /// # extern crate vek;
             /// # #[macro_use] extern crate approx;
@@ -2419,7 +2432,7 @@ macro_rules! mat_impl_mat4 {
             /// Builds a "look at" model transform for right-handed spaces
             /// from an eye position, a target position, and up vector.
             /// Preferred for transforming objects.
-            /// 
+            ///
             /// ```
             /// # extern crate vek;
             /// # #[macro_use] extern crate approx;
@@ -2610,7 +2623,7 @@ macro_rules! mat_impl_mat4 {
             }
 
             /// Creates a perspective projection matrix for right-handed spaces, with zero-to-one depth clip planes.
-            pub fn perspective_rh_zo (fov_y_radians: T, aspect_ratio: T, near: T, far: T) -> Self 
+            pub fn perspective_rh_zo (fov_y_radians: T, aspect_ratio: T, near: T, far: T) -> Self
                 where T: Real + FloatConst + Debug
             {
                 // Assertions from cgmath
@@ -2635,7 +2648,7 @@ macro_rules! mat_impl_mat4 {
                 )
             }
             /// Creates a perspective projection matrix for left-handed spaces, with zero-to-one depth clip planes.
-            pub fn perspective_lh_zo (fov_y_radians: T, aspect_ratio: T, near: T, far: T) -> Self 
+            pub fn perspective_lh_zo (fov_y_radians: T, aspect_ratio: T, near: T, far: T) -> Self
                 where T: Real + FloatConst + Debug
             {
                 let mut m = Self::perspective_rh_zo(fov_y_radians, aspect_ratio, near, far);
@@ -2645,7 +2658,7 @@ macro_rules! mat_impl_mat4 {
             }
 
             /// Creates a perspective projection matrix for right-handed spaces, with negative-one-to-one depth clip planes.
-            pub fn perspective_rh_no (fov_y_radians: T, aspect_ratio: T, near: T, far: T) -> Self 
+            pub fn perspective_rh_no (fov_y_radians: T, aspect_ratio: T, near: T, far: T) -> Self
                 where T: Real + FloatConst + Debug
             {
                 // Assertions from cgmath
@@ -2670,7 +2683,7 @@ macro_rules! mat_impl_mat4 {
                 )
             }
             /// Creates a perspective projection matrix for left-handed spaces, with negative-one-to-one depth clip planes.
-            pub fn perspective_lh_no (fov_y_radians: T, aspect_ratio: T, near: T, far: T) -> Self 
+            pub fn perspective_lh_no (fov_y_radians: T, aspect_ratio: T, near: T, far: T) -> Self
                 where T: Real + FloatConst + Debug
             {
                 let mut m = Self::perspective_rh_no(fov_y_radians, aspect_ratio, near, far);
@@ -2683,7 +2696,7 @@ macro_rules! mat_impl_mat4 {
             ///
             /// # Panics
             /// `width`, `height` and `fov_y_radians` must all be strictly greater than zero.
-            pub fn perspective_fov_rh_zo (fov_y_radians: T, width: T, height: T, near: T, far: T) -> Self 
+            pub fn perspective_fov_rh_zo (fov_y_radians: T, width: T, height: T, near: T, far: T) -> Self
                 where T: Real + FloatConst + Debug
             {
                 debug_assert!(width > T::zero(), "viewport width cannot be below zero, found: {:?}", width);
@@ -2717,7 +2730,7 @@ macro_rules! mat_impl_mat4 {
             ///
             /// # Panics
             /// `width`, `height` and `fov_y_radians` must all be strictly greater than zero.
-            pub fn perspective_fov_lh_zo (fov_y_radians: T, width: T, height: T, near: T, far: T) -> Self 
+            pub fn perspective_fov_lh_zo (fov_y_radians: T, width: T, height: T, near: T, far: T) -> Self
                 where T: Real + FloatConst + Debug
             {
                 let mut m = Self::perspective_fov_rh_zo(fov_y_radians, width, height, near, far);
@@ -2730,7 +2743,7 @@ macro_rules! mat_impl_mat4 {
             ///
             /// # Panics
             /// `width`, `height` and `fov_y_radians` must all be strictly greater than zero.
-            pub fn perspective_fov_rh_no (fov_y_radians: T, width: T, height: T, near: T, far: T) -> Self 
+            pub fn perspective_fov_rh_no (fov_y_radians: T, width: T, height: T, near: T, far: T) -> Self
                 where T: Real + FloatConst + Debug
             {
                 debug_assert!(width > T::zero(), "viewport width cannot be below zero, found: {:?}", width);
@@ -2763,7 +2776,7 @@ macro_rules! mat_impl_mat4 {
             ///
             /// # Panics
             /// `width`, `height` and `fov_y_radians` must all be strictly greater than zero.
-            pub fn perspective_fov_lh_no (fov_y_radians: T, width: T, height: T, near: T, far: T) -> Self 
+            pub fn perspective_fov_lh_no (fov_y_radians: T, width: T, height: T, near: T, far: T) -> Self
                 where T: Real + FloatConst + Debug
             {
                 let mut m = Self::perspective_fov_rh_no(fov_y_radians, width, height, near, far);
@@ -2777,7 +2790,7 @@ macro_rules! mat_impl_mat4 {
             ///
             /// [Link to PDF](http://www.terathon.com/gdc07_lengyel.pdf)
             // From GLM
-            pub fn tweaked_infinite_perspective_rh (fov_y_radians: T, aspect_ratio: T, near: T, epsilon: T) -> Self 
+            pub fn tweaked_infinite_perspective_rh (fov_y_radians: T, aspect_ratio: T, near: T, epsilon: T) -> Self
                 where T: Real + FloatConst + Debug
             {
                 // Assertions from cgmath
@@ -2807,7 +2820,7 @@ macro_rules! mat_impl_mat4 {
             }
 
             /// Creates an infinite perspective projection matrix for left-handed spaces.
-            pub fn tweaked_infinite_perspective_lh (fov_y_radians: T, aspect_ratio: T, near: T, epsilon: T) -> Self 
+            pub fn tweaked_infinite_perspective_lh (fov_y_radians: T, aspect_ratio: T, near: T, epsilon: T) -> Self
                 where T: Real + FloatConst + Debug
             {
                 let mut m = Self::tweaked_infinite_perspective_rh(fov_y_radians, aspect_ratio, near, epsilon);
@@ -2817,14 +2830,14 @@ macro_rules! mat_impl_mat4 {
             }
 
             /// Creates an infinite perspective projection matrix for right-handed spaces.
-            pub fn infinite_perspective_rh (fov_y_radians: T, aspect_ratio: T, near: T) -> Self 
+            pub fn infinite_perspective_rh (fov_y_radians: T, aspect_ratio: T, near: T) -> Self
                 where T: Real + FloatConst + Debug
             {
                 Self::tweaked_infinite_perspective_rh(fov_y_radians, aspect_ratio, near, T::zero())
             }
 
             /// Creates an infinite perspective projection matrix for left-handed spaces.
-            pub fn infinite_perspective_lh (fov_y_radians: T, aspect_ratio: T, near: T) -> Self 
+            pub fn infinite_perspective_lh (fov_y_radians: T, aspect_ratio: T, near: T) -> Self
                 where T: Real + FloatConst + Debug
             {
                 Self::tweaked_infinite_perspective_lh(fov_y_radians, aspect_ratio, near, T::zero())
@@ -3096,7 +3109,7 @@ macro_rules! mat_impl_mat4 {
                 let one = T::one();
                 let two = one + one;
                 let four = two + two;
-                if tr > T::zero() { 
+                if tr > T::zero() {
                     let s = (tr+one).sqrt() * two; // S=4*qw
                     Self {
                         w: s / four,
@@ -3104,15 +3117,15 @@ macro_rules! mat_impl_mat4 {
                         y: (m02 - m20) / s,
                         z: (m10 - m01) / s,
                     }
-                } else if (m00 > m11) && (m00 > m22) { 
-                    let s = (one + m00 - m11 - m22).sqrt() * two; // S=4*qx 
+                } else if (m00 > m11) && (m00 > m22) {
+                    let s = (one + m00 - m11 - m22).sqrt() * two; // S=4*qx
                     Self {
                         w: (m21 - m12) / s,
                         x: s / four,
                         y: (m01 + m10) / s,
                         z: (m02 + m20) / s,
                     }
-                } else if m11 > m22 { 
+                } else if m11 > m22 {
                     let s = (one + m11 - m00 - m22).sqrt() * two; // S=4*qy
                     Self {
                         w: (m02 - m20) / s,
@@ -3120,7 +3133,7 @@ macro_rules! mat_impl_mat4 {
                         y: s / four,
                         z: (m12 + m21) / s,
                     }
-                } else { 
+                } else {
                     let s = (one + m22 - m00 - m11).sqrt() * two; // S=4*qz
                     Self {
                         w: (m10 - m01) / s,
@@ -3277,7 +3290,7 @@ macro_rules! mat_impl_mat3 {
             /// let t = Mat3::new(
             ///     0, 4, 8,
             ///     1, 5, 9,
-            ///     2, 6, 0 
+            ///     2, 6, 0
             /// );
             /// assert_eq!(m.transposed(), t);
             /// assert_eq!(m, m.transposed().transposed());
@@ -3465,21 +3478,21 @@ macro_rules! mat_impl_mat3 {
                 )
             }
 
-            /// Rotates this matrix around a 3D axis.  
+            /// Rotates this matrix around a 3D axis.
             /// The axis is not required to be normalized.
             pub fn rotate_3d<V: Into<Vec3<T>>>(&mut self, angle_radians: T, axis: V)
                 where T: Real + MulAdd<T,T,Output=T> + Sum
             {
                 *self = self.rotated_3d(angle_radians, axis);
             }
-            /// Returns this matrix rotated around a 3D axis.  
+            /// Returns this matrix rotated around a 3D axis.
             /// The axis is not required to be normalized.
             pub fn rotated_3d<V: Into<Vec3<T>>>(self, angle_radians: T, axis: V) -> Self
                 where T: Real + MulAdd<T,T,Output=T> + Sum
             {
                 Self::rotation_3d(angle_radians, axis) * self
             }
-            /// Creates a matrix that rotates around a 3D axis.  
+            /// Creates a matrix that rotates around a 3D axis.
             /// The axis is not required to be normalized.
             ///
             /// ```
@@ -3534,7 +3547,7 @@ macro_rules! mat_impl_mat3 {
                 Self::new(
                     oc*x*x + c  , oc*x*y - z*s, oc*z*x + y*s,
                     oc*x*y + z*s, oc*y*y + c  , oc*y*z - x*s,
-                    oc*z*x - y*s, oc*y*z + x*s, oc*z*z + c  
+                    oc*z*x - y*s, oc*y*z + x*s, oc*z*z + c
                 )
             }
             /// Creates a matrix that would rotate a `from` direction to `to`.
@@ -3554,7 +3567,7 @@ macro_rules! mat_impl_mat3 {
             /// assert_relative_eq!(m * from, to);
             /// # }
             /// ```
-            pub fn rotation_from_to_3d<V: Into<Vec3<T>>>(from: V, to: V) -> Self 
+            pub fn rotation_from_to_3d<V: Into<Vec3<T>>>(from: V, to: V) -> Self
                 where T: Real + Sum
             {
                 Self::from(Quaternion::rotation_from_to_3d(from, to))
@@ -3587,7 +3600,7 @@ macro_rules! mat_impl_mat3 {
                 }
             }
         }
-        impl<T> From<Quaternion<T>> for Mat3<T> 
+        impl<T> From<Quaternion<T>> for Mat3<T>
             where T: Copy + Zero + One + Mul<Output=T> + Add<Output=T> + Sub<Output=T>
         {
             fn from(q: Quaternion<T>) -> Self {
@@ -3608,7 +3621,7 @@ macro_rules! mat_impl_mat3 {
         /// assert_relative_eq!(a, b);
         /// # }
         /// ```
-        impl<T> From<Mat3<T>> for Quaternion<T> 
+        impl<T> From<Mat3<T>> for Quaternion<T>
             where T: Zero + One + Mul<Output=T> + Add<Output=T> + Sub<Output=T> + Real
         {
             fn from(m: Mat3<T>) -> Self {
@@ -3975,8 +3988,8 @@ macro_rules! mat_declare_modules {
             //! Matrices stored in column-major layout.
             //!
             //! Multiplying a column-major matrix by one or more column vectors is fast
-            //! due to the way it is implemented in SIMD 
-            //! (a `matrix * vector` mutliply is four broadcasts, one SIMD product and three fused-multiply-adds).  
+            //! due to the way it is implemented in SIMD
+            //! (a `matrix * vector` mutliply is four broadcasts, one SIMD product and three fused-multiply-adds).
             //!
             //! Because `matrix * matrix` and `matrix * vector` products are fastest with this layout,
             //! it's the preferred one for most computer graphics libraries and application.
@@ -4006,7 +4019,7 @@ macro_rules! mat_declare_modules {
 
 pub mod repr_c {
     //! Matrix types which use `#[repr(C)]` vectors exclusively.
-    //! 
+    //!
     //! See also the `repr_simd` neighbour module, which is available on Nightly
     //! with the `repr_simd` feature enabled.
 
@@ -4026,7 +4039,7 @@ pub mod repr_c {
 #[cfg(all(nightly, feature="repr_simd"))]
 pub mod repr_simd {
     //! Matrix types which use a `#[repr(C)]` vector of `#[repr(simd)]` vectors.
-   
+
     use super::*;
     #[allow(unused_imports)]
     use super::vec::repr_simd::{Vec2};
@@ -4171,7 +4184,7 @@ mod tests {
         let model = Mat4::<f32>::model_look_at_rh(eye, target, Vec4::up());
         assert_relative_eq!(model * Vec4::unit_w(), Vec4::new(1_f32, 0., 1., 1.));
         assert_relative_eq!(model * Vec4::new(0_f32, 0., -2_f32.sqrt(), 1.), Vec4::new(2_f32, 0., 2., 1.));
-        
+
         // A "model" look-at essentially undoes a "view" look-at
         let view = Mat4::<f32>::look_at_rh(eye, target, Vec4::up());
         assert_relative_eq!(view * model, Mat4::identity());
@@ -4186,7 +4199,7 @@ mod tests {
         assert_relative_eq!(model * Vec4::unit_w(), eye);
         let d = 2_f32.sqrt();
         assert_relative_eq!(model * Vec4::new(0_f32, 0., d, 1.), target);
-        
+
         // A "model" look-at essentially undoes a "view" look-at
         let view = Mat4::look_at_lh(eye, target, Vec4::up());
         assert_relative_eq!(view * model, Mat4::identity());

--- a/src/ops.rs
+++ b/src/ops.rs
@@ -209,7 +209,7 @@ impl_muladd!{integer
 
 /// A value that can be linearly interpolated.
 ///
-/// Note that, like standard operators, this can be implement for `T` and `&T`.  
+/// Note that, like standard operators, this can be implement for `T` and `&T`.
 /// You would make the difference like so:
 ///
 /// ```
@@ -224,15 +224,15 @@ impl_muladd!{integer
 /// assert_eq!(a, d);
 /// ```
 ///
-/// This is made possible thanks to the explicit `Output` type.  
+/// This is made possible thanks to the explicit `Output` type.
 /// Therefore, it's also convenient for `GameState` structures, which you might
-/// prefer to interpolate by reference instead of consuming them.  
+/// prefer to interpolate by reference instead of consuming them.
 /// The interpolation of two `&GameState`s would produce a new `GameState` value.
 ///
 /// ```
 /// use vek::{Lerp, Vec3};
-/// 
-/// /// A data-heavy structure that represents a current game state.  
+///
+/// /// A data-heavy structure that represents a current game state.
 /// /// It's neither Copy and nor even Clone!
 /// struct GameState {
 ///     pub camera_position: Vec3<f32>,
@@ -263,7 +263,7 @@ pub trait Lerp<Factor=f32>: Sized
     ///
     /// This would make use of inclusive ranges, but they aren't stable yet.
     ///
-    /// A possible implementation is `from + factor * (to - from)`, a.k.a 
+    /// A possible implementation is `from + factor * (to - from)`, a.k.a
     /// `factor.mul_add(to - from, from)`.
     ///
     /// ```
@@ -367,19 +367,19 @@ macro_rules! lerp_impl_integer {
             impl Lerp<f32> for $T {
                 type Output = $T;
                 fn lerp_unclamped_precise(from: Self, to: Self, factor: f32) -> Self {
-                    ((from as f32)*((1f32)-factor) + (to as f32)*factor).round() as Self
+                    num_traits::Float::round((from as f32)*((1f32)-factor) + (to as f32)*factor) as Self
                 }
                 fn lerp_unclamped(from: Self, to: Self, factor: f32) -> Self {
-                    factor.mul_add((to - from) as f32, from as f32).round() as Self
+                    num_traits::Float::round(factor.mul_add((to - from) as f32, from as f32)) as Self
                 }
             }
             impl Lerp<f64> for $T {
                 type Output = $T;
                 fn lerp_unclamped_precise(from: Self, to: Self, factor: f64) -> Self {
-                    ((from as f64)*((1f64)-factor) + (to as f64)*factor).round() as Self
+                    num_traits::Float::round((from as f64)*((1f64)-factor) + (to as f64)*factor) as Self
                 }
                 fn lerp_unclamped(from: Self, to: Self, factor: f64) -> Self {
-                    factor.mul_add((to - from) as f64, from as f64).round() as Self
+                    num_traits::Float::round(factor.mul_add((to - from) as f64, from as f64)) as Self
                 }
             }
             impl<'a> Lerp<f32> for &'a $T {
@@ -451,7 +451,7 @@ pub trait Wrap<Bound=Self>: Sized {
     /// # Panics
     /// Panics if `upper <= 0`. Reasons include :
     ///
-    /// - Some types may implement it as `self.wrapped_between(zero, upper)`. 
+    /// - Some types may implement it as `self.wrapped_between(zero, upper)`.
     ///   A negative `upper` would violate the `lower <= upper` requirement in this case;
     /// - On unsigned integers, this just resolves to `self % upper`,
     ///   and integer division by zero is forbidden. Testing for `i==0` incurs unnecessary overhead.
@@ -500,8 +500,8 @@ pub trait Wrap<Bound=Self>: Sized {
     ///
     /// # Panics
     /// Panics if `lower >= upper`. Swap the values yourself if necessary.
-    /// 
-    /// Also panics if `lower < 0` or `upper <= 0`. See `wrapped()` for a rationale.  
+    ///
+    /// Also panics if `lower < 0` or `upper <= 0`. See `wrapped()` for a rationale.
     /// Forcing `lower` and `upper` to be positive allows implementations to be simpler and faster.
     ///
     /// ```
@@ -524,7 +524,7 @@ pub trait Wrap<Bound=Self>: Sized {
     ///
     /// # Panics
     /// Panics if `lower` is greater than `upper`. Swap the values yourself if necessary.
-    fn wrap_between(val: Self, lower: Bound, upper: Bound) -> Self 
+    fn wrap_between(val: Self, lower: Bound, upper: Bound) -> Self
         where Self: Sub<Output=Self> + Add<Output=Self> + From<Bound>,
               Bound: Copy + Sub<Output=Bound> + PartialOrd
     {
@@ -587,7 +587,7 @@ macro_rules! wrap_impl_float {
                 fn wrapped(self, upper: Self) -> Self {
                     assert!(upper > Self::zero());
                     // assert_relative_ne!(upper, Self::zero());
-                    self - (self/upper).floor() * upper
+                    self - num_traits::Float::floor(self/upper) * upper
                 }
                 fn wrapped_between(self, lower: Self, upper: Self) -> Self {
                     assert!(lower < upper);
@@ -602,7 +602,7 @@ macro_rules! wrap_impl_float {
                     // assert_relative_ne!(upper, Self::zero());
                     let t = self.wrapped(upper + upper);
                     let upper = || Self::from(upper);
-                    upper() - (t - upper()).abs()
+                    upper() - num_traits::Float::abs(t - upper())
                 }
             }
         )+

--- a/src/quaternion.rs
+++ b/src/quaternion.rs
@@ -1,16 +1,16 @@
 //! Quaternions are a convenient representation for rotations in 3D spaces.
 
-use num_traits::{Zero, One, real::Real};
-use approx::ApproxEq;
-use std::ops::*;
-use std::iter::Sum;
+use approx::{AbsDiffEq, RelativeEq, UlpsEq};
+use num_traits::{real::Real, One, Zero};
 use ops::*;
+use std::iter::Sum;
+use std::ops::*;
 
 macro_rules! impl_mul_by_vec {
     ($Vec3:ident $Vec4:ident) => {
         /// 3D vectors can be rotated by being premultiplied by a quaternion, **assuming the
         /// quaternion is normalized**.
-        /// On `Vec4`s, the `w` element is preserved, so you can safely rotate 
+        /// On `Vec4`s, the `w` element is preserved, so you can safely rotate
         /// points and directions.
         ///
         /// ```
@@ -166,7 +166,7 @@ macro_rules! quaternion_complete_mod {
             /// **Be careful: since it has a magnitude equal to zero, it is not
             /// valid to use for most operations.**
             pub fn zero() -> Self where T: Zero {
-                Self { 
+                Self {
                     x: T::zero(),
                     y: T::zero(),
                     z: T::zero(),
@@ -193,7 +193,7 @@ macro_rules! quaternion_complete_mod {
             /// # }
             /// ```
             pub fn identity() -> Self where T: Zero + One {
-                Self { 
+                Self {
                     x: T::zero(),
                     y: T::zero(),
                     z: T::zero(),
@@ -252,7 +252,7 @@ macro_rules! quaternion_complete_mod {
                 self.conjugate() / self.into_vec4().magnitude_squared()
             }
             /// Gets the dot product between two quaternions.
-            pub fn dot(self, q: Self) -> T 
+            pub fn dot(self, q: Self) -> T
                 where T: Copy + Sum + Mul<Output=T>
             {
                 self.into_vec4().dot(q.into_vec4())
@@ -288,7 +288,7 @@ macro_rules! quaternion_complete_mod {
             /// assert_relative_eq!(q * from, to);
             /// # }
             /// ```
-            pub fn rotation_from_to_3d<V: Into<Vec3<T>>>(from: V, to: V) -> Self 
+            pub fn rotation_from_to_3d<V: Into<Vec3<T>>>(from: V, to: V) -> Self
                 where T: Real + Sum
             {
                 // From GLM
@@ -308,7 +308,7 @@ macro_rules! quaternion_complete_mod {
                 Self { x, y, z, w }.normalized()
             }
 
-            /// Creates a quaternion from an angle and axis.  
+            /// Creates a quaternion from an angle and axis.
             /// The axis is not required to be normalized.
             pub fn rotation_3d<V: Into<Vec3<T>>>(angle_radians: T, axis: V) -> Self
                 where T: Real + Sum
@@ -331,37 +331,37 @@ macro_rules! quaternion_complete_mod {
             pub fn rotation_z(angle_radians: T) -> Self where T: Real + Sum {
                 Self::rotation_3d(angle_radians, Vec3::unit_z())
             }
-            /// Returns this quaternion rotated around the given axis with given angle.  
+            /// Returns this quaternion rotated around the given axis with given angle.
             /// The axis is not required to be normalized.
             pub fn rotated_3d<V: Into<Vec3<T>>>(self, angle_radians: T, axis: V) -> Self where T: Real + Sum {
                 Self::rotation_3d(angle_radians, axis) * self
             }
-            /// Returns this quaternion rotated around the X axis with given angle.  
+            /// Returns this quaternion rotated around the X axis with given angle.
             pub fn rotated_x(self, angle_radians: T) -> Self where T: Real + Sum {
                 Self::rotation_x(angle_radians) * self
             }
-            /// Returns this quaternion rotated around the Y axis with given angle.  
+            /// Returns this quaternion rotated around the Y axis with given angle.
             pub fn rotated_y(self, angle_radians: T) -> Self where T: Real + Sum {
                 Self::rotation_y(angle_radians) * self
             }
-            /// Returns this quaternion rotated around the Z axis with given angle.  
+            /// Returns this quaternion rotated around the Z axis with given angle.
             pub fn rotated_z(self, angle_radians: T) -> Self where T: Real + Sum {
                 Self::rotation_z(angle_radians) * self
             }
-            /// Rotates this quaternion around the given axis with given angle.  
+            /// Rotates this quaternion around the given axis with given angle.
             /// The axis is not required to be normalized.
             pub fn rotate_3d<V: Into<Vec3<T>>>(&mut self, angle_radians: T, axis: V) where T: Real + Sum {
                 *self = self.rotated_3d(angle_radians, axis);
             }
-            /// Rotates this quaternion around the X axis with given angle.  
+            /// Rotates this quaternion around the X axis with given angle.
             pub fn rotate_x(&mut self, angle_radians: T) where T: Real + Sum {
                 *self = self.rotated_x(angle_radians);
             }
-            /// Rotates this quaternion around the Y axis with given angle.  
+            /// Rotates this quaternion around the Y axis with given angle.
             pub fn rotate_y(&mut self, angle_radians: T) where T: Real + Sum {
                 *self = self.rotated_y(angle_radians);
             }
-            /// Rotates this quaternion around the Z axis with given angle.  
+            /// Rotates this quaternion around the Z axis with given angle.
             pub fn rotate_z(&mut self, angle_radians: T) where T: Real + Sum {
                 *self = self.rotated_z(angle_radians);
             }
@@ -408,7 +408,7 @@ macro_rules! quaternion_complete_mod {
             pub fn into_vec4(self) -> Vec4<T> {
                 self.into()
             }
-            /// Creates a quaternion from a `Vec4` by destructuring.  
+            /// Creates a quaternion from a `Vec4` by destructuring.
             /// **You are responsible for ensuring that the resulting quaternion is normalized.**
             pub fn from_vec4(v: Vec4<T>) -> Self {
                 v.into()
@@ -434,7 +434,7 @@ macro_rules! quaternion_complete_mod {
                 mint::Quaternion { s, v: v.into() }
             }
         }
-        
+
         /// The `Lerp` implementation for quaternion is the "Normalized LERP".
         impl<T, Factor> Lerp<Factor> for Quaternion<T>
             where T: Lerp<Factor,Output=T> + Sum + Real,
@@ -464,13 +464,13 @@ macro_rules! quaternion_complete_mod {
                 Lerp::lerp_unclamped(*from, *to, factor)
             }
         }
-        impl<T> Quaternion<T> 
+        impl<T> Quaternion<T>
             where T: Copy + One + Mul<Output=T> + Sub<Output=T> + MulAdd<T,T,Output=T>
         {
             /// Performs linear interpolation **without normalizing the result**,
             /// using an implementation that supposedly yields a more precise result.
-            /// 
-            /// This is probably not what you're looking for.  
+            ///
+            /// This is probably not what you're looking for.
             /// For an implementation that normalizes the result (which is more commonly wanted), see the `Lerp` implementation.
             pub fn lerp_precise_unnormalized(from: Self, to: Self, factor: T) -> Self where T: Clamp + Zero {
                 Self::lerp_unclamped_precise_unnormalized(from, to, factor.clamped01())
@@ -478,27 +478,27 @@ macro_rules! quaternion_complete_mod {
             /// Performs linear interpolation **without normalizing the result** and without
             /// implicitly constraining `factor` to be between 0 and 1,
             /// using an implementation that supposedly yields a more precise result.
-            /// 
-            /// This is probably not what you're looking for.  
+            ///
+            /// This is probably not what you're looking for.
             /// For an implementation that normalizes the result (which is more commonly wanted), see the `Lerp` implementation.
             pub fn lerp_unclamped_precise_unnormalized(from: Self, to: Self, factor: T) -> Self {
                 Vec4::lerp_unclamped_precise(from.into(), to.into(), factor).into()
             }
         }
-        impl<T> Quaternion<T> 
+        impl<T> Quaternion<T>
             where T: Copy + Sub<Output=T> + MulAdd<T,T,Output=T>
         {
             /// Performs linear interpolation **without normalizing the result**.
-            /// 
-            /// This is probably not what you're looking for.  
+            ///
+            /// This is probably not what you're looking for.
             /// For an implementation that normalizes the result (which is more commonly wanted), see the `Lerp` implementation.
             pub fn lerp_unnormalized(from: Self, to: Self, factor: T) -> Self where T: Clamp + Zero + One {
                 Self::lerp_unclamped_unnormalized(from, to, factor.clamped01())
             }
             /// Performs linear interpolation **without normalizing the result** and without
             /// implicitly constraining `factor` to be between 0 and 1.
-            /// 
-            /// This is probably not what you're looking for.  
+            ///
+            /// This is probably not what you're looking for.
             /// For an implementation that normalizes the result (which is more commonly wanted), see the `Lerp` implementation.
             pub fn lerp_unclamped_unnormalized(from: Self, to: Self, factor: T) -> Self {
                 Vec4::lerp_unclamped(from.into(), to.into(), factor).into()
@@ -519,7 +519,7 @@ macro_rules! quaternion_complete_mod {
             /// # fn main() {
             /// let from = Quaternion::rotation_z(0_f32);
             /// let to = Quaternion::rotation_z(PI*9./10.);
-            /// 
+            ///
             /// let angles = 32;
             /// for i in 0..angles {
             ///     let factor = (i as f32) / (angles as f32);
@@ -532,7 +532,7 @@ macro_rules! quaternion_complete_mod {
             // From GLM's source code.
             pub fn slerp_unclamped(from: Self, mut to: Self, factor: T) -> Self {
                 let mut cos_theta = from.dot(to);
-                // If cosTheta < 0, the interpolation will take the long way around the sphere. 
+                // If cosTheta < 0, the interpolation will take the long way around the sphere.
                 // To fix this, one quat must be negated.
                 if cos_theta < T::zero() {
                     to = -to;
@@ -553,7 +553,7 @@ macro_rules! quaternion_complete_mod {
             }
         }
 
-        impl<T, Factor> Slerp<Factor> for Quaternion<T> 
+        impl<T, Factor> Slerp<Factor> for Quaternion<T>
             where T: Lerp<T,Output=T> + Sum + Real,
                   Factor: Into<T>
         {
@@ -562,7 +562,7 @@ macro_rules! quaternion_complete_mod {
                 Self::slerp_unclamped(from, to, factor.into())
             }
         }
-        impl<'a, T, Factor> Slerp<Factor> for &'a Quaternion<T> 
+        impl<'a, T, Factor> Slerp<Factor> for &'a Quaternion<T>
             where T: Lerp<T,Output=T> + Sum + Real,
                   Factor: Into<T>
         {
@@ -637,13 +637,13 @@ macro_rules! quaternion_complete_mod {
         /// assert_relative_eq!(q*p*v, -Vec4::unit_z());
         /// # }
         /// ```
-        impl<T> Mul for Quaternion<T> 
+        impl<T> Mul for Quaternion<T>
             where T: Copy + Mul<Output=T> + Sub<Output=T> + Zero + Sum
         {
             type Output = Self;
             fn mul(self, rhs: Self) -> Self::Output {
                 let ((ps, pv), (qs, qv)) = (
-                    self.into_scalar_and_vec3(), 
+                    self.into_scalar_and_vec3(),
                     rhs.into_scalar_and_vec3()
                 );
                 let Vec3 { x, y, z } = qv*ps + pv*qs + pv.cross(qv);
@@ -679,27 +679,30 @@ macro_rules! quaternion_complete_mod {
         }
         */
 
-
-        impl<T: ApproxEq> ApproxEq for Quaternion<T> where T::Epsilon: Copy {
+        impl<T: AbsDiffEq> AbsDiffEq for Quaternion<T>
+        where
+            T::Epsilon: Copy,
+        {
             type Epsilon = T::Epsilon;
 
             fn default_epsilon() -> T::Epsilon {
                 T::default_epsilon()
             }
 
-            fn default_max_relative() -> T::Epsilon {
-                T::default_max_relative()
+            fn abs_diff_eq(&self, other: &Self, epsilon: Self::Epsilon) -> bool {
+                T::abs_diff_eq(&self.w, &other.w, epsilon)
+                    && T::abs_diff_eq(&self.x, &other.x, epsilon)
+                    && T::abs_diff_eq(&self.y, &other.y, epsilon)
+                    && T::abs_diff_eq(&self.z, &other.z, epsilon)
+            }
             }
 
+        impl<T: UlpsEq> UlpsEq for Quaternion<T>
+        where
+            T::Epsilon: Copy,
+        {
             fn default_max_ulps() -> u32 {
                 T::default_max_ulps()
-            }
-
-            fn relative_eq(&self, other: &Self, epsilon: T::Epsilon, max_relative: T::Epsilon) -> bool {
-                   T::relative_eq(&self.w, &other.w, epsilon, max_relative)
-                && T::relative_eq(&self.x, &other.x, epsilon, max_relative)
-                && T::relative_eq(&self.y, &other.y, epsilon, max_relative)
-                && T::relative_eq(&self.z, &other.z, epsilon, max_relative)
             }
 
             fn ulps_eq(&self, other: &Self, epsilon: T::Epsilon, max_ulps: u32) -> bool {
@@ -709,7 +712,28 @@ macro_rules! quaternion_complete_mod {
                 && T::ulps_eq(&self.z, &other.z, epsilon, max_ulps)
             }
         }
+
+        impl<T: RelativeEq> RelativeEq for Quaternion<T>
+        where
+            T::Epsilon: Copy,
+        {
+            fn default_max_relative() -> T::Epsilon {
+                T::default_max_relative()
     }
+
+            fn relative_eq(
+                &self,
+                other: &Self,
+                epsilon: T::Epsilon,
+                max_relative: T::Epsilon,
+            ) -> bool {
+                T::relative_eq(&self.w, &other.w, epsilon, max_relative)
+                    && T::relative_eq(&self.x, &other.x, epsilon, max_relative)
+                    && T::relative_eq(&self.y, &other.y, epsilon, max_relative)
+                    && T::relative_eq(&self.z, &other.z, epsilon, max_relative)
+}
+    }
+    };
 }
 
 
@@ -752,7 +776,7 @@ mod tests {
         #[test] fn rotation_from_to_3d() {
             let a = Vec3::new(1_f32, 200., 3.);
             let b = Vec3::new(80_f32, 0., 352.);
-            let q = Quaternion::rotation_from_to_3d(a, b);
+            let q = Quaternion::<f32>::rotation_from_to_3d(a, b);
             assert_relative_eq!(q.magnitude(), 1.);
         }
         #[test] fn rotation_3d() {

--- a/src/vec.rs
+++ b/src/vec.rs
@@ -14,7 +14,7 @@ use std::cmp;
 use std::ops::*;
 use std::slice::{self, /*SliceIndex*/}; // NOTE: Will want to use SliceIndex once it's stabilized
 use num_traits::{Zero, One, NumCast, Signed, real::Real};
-use approx::ApproxEq;
+use approx::{AbsDiffEq, RelativeEq, UlpsEq};
 use ops::*;
 
 macro_rules! vec_impl_cmp {
@@ -1177,34 +1177,46 @@ macro_rules! vec_impl_vec {
         // - fn classify(self) -> FpCategory;
         // - fn integer_decode(self) -> (u64, i16, i8);
 
-
-        impl<T: ApproxEq> ApproxEq for $Vec<T> where T::Epsilon: Copy {
+        impl<T: AbsDiffEq> AbsDiffEq for $Vec<T> where T::Epsilon: Copy {
             type Epsilon = T::Epsilon;
 
             fn default_epsilon() -> T::Epsilon {
                 T::default_epsilon()
             }
 
-            fn default_max_relative() -> T::Epsilon {
-                T::default_max_relative()
-            }
-
-            fn default_max_ulps() -> u32 {
-                T::default_max_ulps()
-            }
-
-            fn relative_eq(&self, other: &Self, epsilon: T::Epsilon, max_relative: T::Epsilon) -> bool {
+            fn abs_diff_eq(&self, other: &Self, epsilon: Self::Epsilon) -> bool {
                 for (l, r) in self.iter().zip(other.iter()) {
-                    if !T::relative_eq(l, r, epsilon, max_relative) {
+                    if !T::abs_diff_eq(l, r, epsilon) {
                         return false;
                     }
                 }
                 true
             }
+        }
+
+        impl<T: UlpsEq> UlpsEq for $Vec<T> where T::Epsilon: Copy {
+            fn default_max_ulps() -> u32 {
+                T::default_max_ulps()
+            }
 
             fn ulps_eq(&self, other: &Self, epsilon: T::Epsilon, max_ulps: u32) -> bool {
                 for (l, r) in self.iter().zip(other.iter()) {
                     if !T::ulps_eq(l, r, epsilon, max_ulps) {
+                        return false;
+                    }
+                }
+                true
+            }
+        }
+
+        impl<T: RelativeEq> RelativeEq for $Vec<T> where T::Epsilon: Copy {
+            fn default_max_relative() -> T::Epsilon {
+                T::default_max_relative()
+            }
+
+            fn relative_eq(&self, other: &Self, epsilon: T::Epsilon, max_relative: T::Epsilon) -> bool {
+                for (l, r) in self.iter().zip(other.iter()) {
+                    if !T::relative_eq(l, r, epsilon, max_relative) {
                         return false;
                     }
                 }
@@ -1546,8 +1558,8 @@ macro_rules! vec_impl_spatial {
                 self / self.magnitude()
             }
             /// Get a copy of this direction vector such that its length equals 1.
-            /// If all components approximately zero, None is returned (uses ApproxEq).
-            pub fn try_normalized(self) -> Option<Self> where T: ApproxEq + Sum + Real {
+            /// If all components approximately zero, None is returned (uses RelativeEq).
+            pub fn try_normalized(self) -> Option<Self> where T: RelativeEq + Sum + Real {
                 if self.is_approx_zero() {
                     None
                 } else {
@@ -1558,12 +1570,12 @@ macro_rules! vec_impl_spatial {
             pub fn normalize(&mut self) where T: Sum + Real {
                 *self = self.normalized();
             }
-            /// Is this vector normalized ? (Uses `ApproxEq`)
-            pub fn is_normalized(self) -> bool where T: ApproxEq + Sum + Real {
+            /// Is this vector normalized ? (Uses `RelativeEq`)
+            pub fn is_normalized(self) -> bool where T: RelativeEq + Sum + Real {
                 self.magnitude_squared().relative_eq(&T::one(), T::default_epsilon(), T::default_max_relative())
             }
-            /// Is this vector approximately zero ? (Uses `ApproxEq`)
-            pub fn is_approx_zero(self) -> bool where T: ApproxEq + Sum + Real {
+            /// Is this vector approximately zero ? (Uses `RelativeEq`)
+            pub fn is_approx_zero(self) -> bool where T: RelativeEq + Sum + Real {
                 self.map(|n| n.relative_eq(&T::zero(), T::default_epsilon(), T::default_max_relative())).reduce_and()
             }
             /// Get the smallest angle, in radians, between two direction vectors.
@@ -1933,20 +1945,20 @@ macro_rules! vec_impl_spatial_4d {
                 }
                 /// Returns true if this vector is homogeneous (`w = 0` or `w = 1`).
                 ///
-                /// Uses `ApproxEq`.
-                pub fn is_homogeneous(self) -> bool where T: ApproxEq + Zero + One + Copy {
+                /// Uses `RelativeEq`.
+                pub fn is_homogeneous(self) -> bool where T: RelativeEq + Zero + One + Copy {
                     self.is_point() || self.is_direction()
                 }
                 /// Returns true if this vector is a homogeneous point (`w = 1`).
                 ///
-                /// Uses `ApproxEq`.
-                pub fn is_point(self) -> bool where T: ApproxEq + One {
+                /// Uses `RelativeEq`.
+                pub fn is_point(self) -> bool where T: RelativeEq + One {
                     self.w.relative_eq(&T::one(), T::default_epsilon(), T::default_max_relative())
                 }
                 /// Returns true if this vector is a homogeneous direction (`w = 0`).
                 ///
-                /// Uses `ApproxEq`.
-                pub fn is_direction(self) -> bool where T: ApproxEq + Zero {
+                /// Uses `RelativeEq`.
+                pub fn is_direction(self) -> bool where T: RelativeEq + Zero {
                     self.w.relative_eq(&T::zero(), T::default_epsilon(), T::default_max_relative())
                 }
             }

